src/smtsolvers/CoreSMTSolver.cc

```diff
@@ -1364,7 +1387,118 @@ void CoreSMTSolver::learntSizeAdjust() {
     }
 }
 
+bool CoreSMTSolver::vivify_if_needed()
+{
+    if (conflicts < next_vivify) {
+        return ok;
+    }
+    next_vivify = conflicts*1.1 + 30000;
+    vivif_lit_rem = 0;
+    vivif_cl_rem = 0;
+    uint32_t vivif_tried = 0;
+
+    release_assert(decisionLevel() == 0);
+    release_assert(ok);
+    for (auto & vd: vardata) {
+        vd.reason = CRef_Undef;
+    }
+
+    uint32_t j = 0;
+    uint32_t i = 0;
+    for (; i < learnts.size(); i++) {
+        CRef offs = learnts[i];
+        if (!ok) {
+            learnts[j++] = offs;
+            continue;
+        }
+        release_assert(trail.size() == qhead);
+
+        Clause & cl = ca[offs];
+        if (cl.getGlue() > 2 || cl.isVivified()) {
+            learnts[j++] = offs;
+            continue;
+        }
+
+        vivif_tried++;
+        cl.setVivified();
+        if (vivify_one_clause(cl, offs)) {
+            learnts[j++] = offs;
+        } else {
+            ca.free(offs);
+        }
+    }
+    learnts.shrink(i-j);
+
+    if (verbosity) {
+        std::cout << "; Vivification finished."
+                  << " Tried: " << vivif_tried
+                  << " Lit rem: " << vivif_lit_rem << std::endl;
+    }
+
+    return ok;
+}
+
+
+// returns TRUE if we need to keep the clause
+bool CoreSMTSolver::vivify_one_clause(Clause& cl, CRef offs)
+{
+    release_assert(cl.size() > 0);
+    release_assert(decisionLevel() == 0);
+    release_assert(ok);
+
+    vivif_tmp_lits.clear();
+    vivif_tmp_lits.insert(vivif_tmp_lits.end(), cl.begin(), cl.end());
+
+    //std::random_shuffle(vivif_tmp_lits.begin(), vivif_tmp_lits.end());
+    vivif_new.clear();
 
+    newDecisionLevel();
+
+    for (Lit l : vivif_tmp_lits) {
+        if (value(l) == l_True) {
+            cancelUntil(0);
+            detachClause(offs);
+            return false;
```

Comment:
I switched this from `true` to `false`, since it made sense to me that the clause would also be removed from the clause database if it gets detached.